webapi/src/main/scala/org/knora/webapi/responders/v2/ResourceUtilV2.scala

```diff
@@ -259,30 +211,35 @@ final case class ResourceUtilV2Live(triplestore: TriplestoreService, messageRela
     valueContents: Seq[FileValueContentV2],
     requestingUser: User,
   ): Task[T] = {
-    val valuesToConsider = valueContents.filter(!_.isInstanceOf[StillImageExternalFileValueContentV2])
-    // Was this update a success?
+    val temporaryFiles = valueContents.filterNot(_.is[StillImageExternalFileValueContentV2])
     updateTask.foldZIO(
       (e: Throwable) => {
-        // The update failed. Ask Sipi to delete the temporary files and return the original failure.
-        val deleteRequests =
-          valuesToConsider.map(value => DeleteTemporaryFileRequest(value.fileValue.internalFilename, requestingUser))
-        val sipiResults = ZIO.foreachDiscard(deleteRequests)(messageRelay.ask[SuccessResponseV2](_).logError)
-        sipiResults.ignore *> ZIO.fail(e)
+        ZIO
+          .foreachDiscard(temporaryFiles) { file =>
+            sipiService
+              .deleteTemporaryFile(DeleteTemporaryFileRequest(file.fileValue.internalFilename, requestingUser))
+              .logError
+          }
+          .ignore *> ZIO.fail(e)
       },
       (updateInProject: T) => {
-        // Yes. Ask Sipi to move the file to permanent storage.
-        // If Sipi succeeds, return the original result.
-        // Otherwise, return the failures from Sipi.
-        val moveRequests = valuesToConsider
-          .map(_.fileValue.internalFilename)
-          .map(MoveTemporaryFileToPermanentStorageRequest(_, updateInProject.projectADM.shortcode, requestingUser))
-        ZIO.foreachDiscard(moveRequests)(messageRelay.ask[SuccessResponseV2](_)).as(updateInProject)
+        ZIO
+          .foreachDiscard(temporaryFiles) { file =>
+            sipiService.moveTemporaryFileToPermanentStorage(
+              MoveTemporaryFileToPermanentStorageRequest(
+                file.fileValue.internalFilename,
+                updateInProject.projectADM.shortcode,
+                requestingUser,
+              ),
+            )
+          }
+          .as(updateInProject)
       },
     )
   }
 }
 
 object ResourceUtilV2Live {
-  val layer: URLayer[TriplestoreService & MessageRelay, ResourceUtilV2] =
-    ZLayer.fromFunction(ResourceUtilV2Live.apply _)
+  val layer: URLayer[TriplestoreService & SipiService, ResourceUtilV2] =
+    ZLayer.fromFunction(ResourceUtilV2Live.apply(_, _))
```

Comment:
comment: maybe not something you want to play with after your trip to ZIO DI hell... but I think this could be simplified with `ZLayer.derive[ResourceUtilV2Live]`

Reply:
✔️